Implement automatic sheet loading

Esper89 opened the pull request:
I've implemented (and tested) the main feature planned for this mod. The mod now automatically loads language sheets for all mods from their `{modDir}/languages/{lang}.json` to a sheet at `Mods.{modId}`.

For example, if you have the following file structure:

```
BepInEx/
    plugins/
        Silksong.I18N/
            Silksong.I18N.dll
        Someone.SomeOtherMod/
            Someone.SomeOtherMod.dll
            languages/
                en.json
                fr.json
```

Then whenever the game is in English, the contents of `en.json` will be loaded as a sheet named `Mods.Someone.SomeOtherMod`.

A fallback language can be set in your project file using the `NeutralLanguage` property. If `Someone.SomeOtherMod.csproj` contains the following:

```xml
<PropertyGroup>
    ...
    <NeutralLanguage>EN</NeutralLanguage>
    ...
</PropertyGroup>
```

Then whenever the game is using a language other than English, it will default to using the text in `en.json` if a translation is missing. This feature works even if the game is in French and `fr.json` is an incomplete translation—the data in `fr.json` is merged on top of the data in `en.json`, overriding it only when it exists.

The JSON files themselves are written like so:

```json
{
    "SOME_TEXT": "This sure is some text!",
    "NEW_TOOL_NAME": "Super Cool Tool",
    "NEW_TOOL_DESC": "It rhymes!!!",
    "AND_SOME_DIALOGUE": "Hi! I’m an NPC…<hpage>And I’m Hornet.<page>Dialogue tags don’t need to be escaped here."
}
```

And if that were the contents of `en.json`, the text `Super Cool Tool` would now be accessible at the following location whenever the game is in English:

```cs
new TeamCherry.Localization.LocalisedString("Mods.Someone.SomeOtherMod", "NEW_TOOL_NAME")`
```

All of the files and directories involved are loaded case-insensitively, so if a mod works on Windows it will work on Unix and vice-versa. If multiple files match the case-insensitive search (e.g. if `Languages/EN.json` and `lanGUages/en.jSoN` both exist) then the file paths are sorted lexicographically and loaded one after the other, where the contents of files later in the ordering take precedence.

Esper89 commented:
> Do these align nicely with the TC language codes?

They almost definitely do (language codes are *supposed* to be standardized in the first place…) but in this case it doesn't matter—the text can be anything, and the I18N plugin should be the only thing that cares about it anyways.

> …csproj editing is often a tall order.

😅 I didn't know this. We can provide instructions on how to specify a default language from C# as well—it's just an assembly attribute, you can set it like so:

```cs
[assembly: System.Resources.NeutralResourcesLanguage("EN")]
```

DemoJameson commented:
```cs
Dictionary<string, string>? LoadModSheet(
    string modDir,
    string lang,
    Dictionary<string, string>? fallback = null
)
```
When the `lang.json` file does not exist, this method should return `fallback` instead of `null`.

Esper89 commented:
I've cleaned up the plugin declaration (I misunderstood how `BepInAutoPlugin` worked) and the scene loading code (for some reason it didn't work when I first tried patching in `Start`, but it works now) per request.

I also added documentation to the readme explaining how to use this system and how users can translate mods using this system. The documentation would need to be updated if this mod is published on Thunderstore to include details on how to specify this mod as a Thunderstore dependency.

Additionally, if this PR is merged, a feature should be added to [Silksong.Modding.Templates](https://github.com/silksong-modding/Silksong.Modding.Templates) to automatically include `languages/*.json` in the mod's output directory, and then the documentation here should be updated to reflect that.

BadMagic100 commented on README.md:
```diff
@@ -1,3 +1,165 @@
+
+The allowed languages are: `en.json` (English), `fr.json` (French), `de.json` (German), `zh.json`
+(Simplified Chinese), `es.json` (Spanish), `ko.json` (Korean), `ja.json` (Japanese), `it.json`
+(Italian), `pt.json` (Brazillian Portugese), and `ru.json` (Russian). You cannot translate a mod
```
Is this a limitation of implementation or just the fact that TC only supports these languages? Basically wondering if another mod (or tc) added a language would we need a change to this mod to support it? I assume the answer is probably no but just double checking

BadMagic100 commented on README.md:
````diff
@@ -1,3 +1,165 @@
+[BepInDependency("org.silksong-modding.i18n")]
+```
+
+This mod does not need to be added to your project as a reference.
````
But it does need to be added as a thunderstore dependency

Esper89 replied:
I assumed I shouldn't add that now because this mod isn't published on Thunderstore yet, but I can do that.

BadMagic100 commented:
> Additionally, if this PR is merged, a feature should be added to [Silksong.Modding.Templates](https://github.com/silksong-modding/Silksong.Modding.Templates) to automatically include languages/*.json in the mod's output directory, and then the documentation here should be updated to reflect that.

I've been noodling on the best way to do this since yesterday actually, what I'm leaning towards is each core mod getting a feature flag in the template, so if you know what features you need you can just ask for them directly (or even get them all). I think it would probably be difficult to get folders that might not exist in by default in general.

For this pr you may want to consider documenting how to get language files included (based off main branch of template repo). Namely that would be updating the Binaries item list in the copy target. A copy block may also be required in thunderstore.toml

capitalistspz commented on I18NPlugin.cs:
```diff
@@ -1,13 +1,160 @@
+                continue;
+            }
+
+            var modAsm = mod.GetType().Assembly;
```
`GetType().Assembly` returns the BepInEx assembly

BadMagic100 commented on I18NPlugin.cs:
```diff
@@ -1,13 +1,160 @@
+            }
+
+            var isPluginsDir =
+                string.Compare(
```
These would still show up as build warnings afaik

flibber-hk commented on I18NPlugin.cs:
```diff
@@ -1,13 +1,159 @@
+
+    void LoadAllModSheets()
+    {
+        var lang = Language._currentLanguage;
```
Fair enough

BadMagic100 commented on I18NPlugin.cs:
```diff
@@ -1,13 +1,161 @@
+                continue;
+            }
+
+            var isPluginsDir = string.Equals(
```
This could be made a bit cleaner (less verbose + clearer intent) by leveraging DirectoryInfo (.Exists on L53 and .Parent here)

BadMagic100 commented on I18NPlugin.cs:
```diff
@@ -1,13 +1,160 @@
+            if (langAttr is not null)
+            {
+                var fallbackLang = langAttr.CultureName.ToUpper();
+                fallbackSheet = this.LoadModSheet(modDir, fallbackLang.ToLower());
```
I think it's worth adding a comment to this effect (that is, why we are doing it this way).

BadMagic100 commented on README.md:
````diff
@@ -1,3 +1,212 @@
+```
+
+You should edit `Localized text!`, `Super Cool Tool`, `The name of this tool rhymes. How amusing.`,
+and `Hi! I’m an NPC…<hpage>And I’m Hornet.<page>Yay!` by replacing the text with its translation
````
We should probably document the availability and meaning of the special tokens like `<hpage>`, `<page>`, and `<br>`

Esper89 replied:
I don't actually know the specifics of that system, I've barely written any dialogue before. 😅 Probably a good idea to document it but maybe in another PR?

BadMagic100 commented on README.md:
````diff
@@ -1,3 +1,212 @@
+directory by adding `/%(Binaries.Dir)` to its `DestinationFolder` attribute as well, like so:
+
+```xml
+<Copy SourceFiles="@(Binaries)" DestinationFolder="$(ThunderstoreDir)/tmp/%(Binaries.Dir)" />
````
```suggestion
<Copy SourceFiles="@(Binaries)" DestinationFolder="$(ThunderstoreDir)/temp/%(Binaries.Dir)" />
```

I had to change this between 2.0.0-alpha.3 and 2.0.0 because of jank uncovered in testing. Doesn't actually break the build or anything but users won't see `tmp` by default anymore is all

BadMagic100 commented on README.md:
```diff
@@ -1,3 +1,212 @@
+
+To access localized text you can use the `TeamCherry.Localization.LocalisedString` and
+`TeamCherry.Localization.Language` classes. Keep in mind that the `LocalisedString` class name uses
+the Australian English spelling (“localised”) but the namespace `TeamCherry.Localization` does not.
```
This is really funny

Esper89 commented on I18NPlugin.cs:
```diff
@@ -154,8 +151,42 @@ void LoadAllModSheets()
+    private static void OnGetLocalizedText(string? key, string? sheetTitle) =>
+        I18NPlugin.Instance?.CheckKeyExists(sheetTitle, key);
+
+#pragma warning disable Harmony003
```
Harmony thinks I'm assigning to `__instance.Sheet` and `__instance.Key` (I'm not) and thinks that if I were assigning to them then `__instance` would need to be a `ref` parameter (it wouldn't).

BadMagic100 commented on README.md:
````diff
@@ -1,3 +1,214 @@
+```
+BepInEx/
+    plugins/
+        Silksong.I18N/
````
```suggestion
        silksong_modding-I18N/
```
